File: backend_kotlin/src/test/kotlin/be/swsb/makeastats/kotlinbackend/services/db/JdbiPreparedEmbeddedPostgres.java
package be.swsb.makeastats.kotlinbackend.services.db;

import com.opentable.db.postgres.embedded.FlywayPreparer;
import com.opentable.db.postgres.junit.EmbeddedPostgresRules;
import com.opentable.db.postgres.junit.PreparedDbRule;
import org.jdbi.v3.core.Jdbi;
import org.jdbi.v3.testing.JdbiRule;
import org.junit.runner.Description;
import org.junit.runners.model.Statement;

public class JdbiPreparedEmbeddedPostgres extends JdbiRule {

    private final PreparedDbRule embeddedPreparedDbRule;

    private JdbiPreparedEmbeddedPostgres(final String... migrationLocations) {
        String[] locations = migrationLocations != null && migrationLocations.length > 0
                                ? migrationLocations
                                : new String[]{"db/migration"};
        embeddedPreparedDbRule = EmbeddedPostgresRules.preparedDatabase(FlywayPreparer.forClasspathLocation(locations));
    }

    public static JdbiPreparedEmbeddedPostgres preparedJdbi(String... migrationLocations){
        return new JdbiPreparedEmbeddedPostgres(migrationLocations);
    }

    @Override
    protected Jdbi createJdbi() {
        return Jdbi.create(embeddedPreparedDbRule.getTestDatabase());
    }

    @Override
    public Statement apply(final Statement base, final Description description) {
        return embeddedPreparedDbRule.apply(super.apply(base, description), description);
    }
}

File: backend_kotlin/application/src/test/kotlin/be/swsb/makeastats/kotlinbackend/test/JdbiPreparedEmbeddedPostgres.kt
package be.swsb.makeastats.kotlinbackend.test

import com.opentable.db.postgres.embedded.FlywayPreparer
import com.opentable.db.postgres.junit.EmbeddedPostgresRules
import com.opentable.db.postgres.junit.PreparedDbRule
import org.jdbi.v3.core.Jdbi
import org.jdbi.v3.testing.JdbiRule
import org.junit.runner.Description
import org.junit.runners.model.Statement

class JdbiPreparedEmbeddedPostgresKotlin private constructor(vararg migrationLocations: String) : JdbiRule() {

    private val embeddedPreparedDbRule: PreparedDbRule

    init {
        val locations = migrationLocations.allOrElse(arrayOf("db/migration"))
        embeddedPreparedDbRule = EmbeddedPostgresRules.preparedDatabase(FlywayPreparer.forClasspathLocation(*locations))
    }

    override fun createJdbi(): Jdbi {
        return Jdbi.create(embeddedPreparedDbRule.testDatabase)
    }

    override fun apply(base: Statement, description: Description?): Statement {
        return embeddedPreparedDbRule.apply(super.apply(base, description), description)
    }

    companion object {
        fun preparedJdbi(vararg migrationLocations: String): JdbiPreparedEmbeddedPostgresKotlin {
            return JdbiPreparedEmbeddedPostgresKotlin(*migrationLocations)
        }
    }
}

fun <T> Array<out T>.allOrElse(defaultArray: Array<T>): Array<out T> {
    return if (size > 0) this else defaultArray
}